### PR TITLE
hilbert: implement paper version of MJ's 1dPart()

### LINKED ISSUE #201: MultiJagged's compute_split_positions does a lot of duplicate work
While measuring strong scaling of some algorithms, HilbertCurve turned out to be a bit too good.

base strong

This might be a bug in part-bench. With `RAYON_NUM_THREADS=1`, 

```
     Running `/tmp/rust-target/release/part-bench -m ../meshes/triangles_large.meshb -w /dev/shm/triangles_large.weights -a hilbert,4096`
 -> Dimension: 2
 -> Number of nodes: 9893828
 -> Number of elements: 19775488
Benchmarking triangles_large;hilbert,4096
Benchmarking triangles_large;hilbert,4096: Warming up for 3.0000 s

Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 1059.5s, or reduce sample count to 10.
Benchmarking triangles_large;hilbert,4096: Collecting 100 samples in estimated 1059.5 s (100 iterations)
```

Adding `-e` to the command execution for strong scaling measurement:

```
     Running `/tmp/rust-etre-rapide-tant-que-ssd-nvme/release/part-bench -m ../meshes/triangles_large.meshb -w /dev/shm/triangles_large.weights -a hilbert,4096 -e`
 -> Dimension: 2
 -> Number of nodes: 9893828
 -> Number of elements: 19775488
Benchmarking triangles_large;hilbert,4096/threads=1
Benchmarking triangles_large;hilbert,4096/threads=1: Warming up for 3.0000 s

Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 4956.3s, or reduce sample count to 10.
Benchmarking triangles_large;hilbert,4096/threads=1: Collecting 100 samples in estimated 4956.3 s (100 iterations)
```

In both cases, HilbertCurve runs on 1 thread, but the second invocation takes 5x more time.

The two differences between both invocations are:

- efficiency measurements are done with threads pinned to cores. However, a patched part-bench which pins threads for other measurements yield the same results as above,
- efficiency measurements aren't done with rayon's global thread pool: this might be a bug in rayon. the following mesh-part patch serves as proof of concept.

```patch
diff --git a/tools/src/bin/mesh-part.rs b/tools/src/bin/mesh-part.rs
index 547ae0c..b318fb4 100644
--- a/tools/src/bin/mesh-part.rs
+++ b/tools/src/bin/mesh-part.rs
@@ -137,10 +137,24 @@ fn main() -> Result<()> {
     let mesh = mesh?;
     let weights = weights?;
 
-    let partition = match mesh.dimension() {
-        2 => main_d::<2>(matches, edge_weights, mesh, weights)?,
-        3 => main_d::<3>(matches, edge_weights, mesh, weights)?,
-        n => anyhow::bail!("expected 2D or 3D mesh, got a {n}D mesh"),
+    let partition = if std::env::var("SLOW").is_ok() {
+        rayon::ThreadPoolBuilder::new()
+            .build()
+            .unwrap()
+            .install(|| {
+                let partition = match mesh.dimension() {
+                    2 => main_d::<2>(matches, edge_weights, mesh, weights)?,
+                    3 => main_d::<3>(matches, edge_weights, mesh, weights)?,
+                    n => anyhow::bail!("expected 2D or 3D mesh, got a {n}D mesh"),
+                };
+                Ok(partition)
+            })?
+    } else {
+        match mesh.dimension() {
+            2 => main_d::<2>(matches, edge_weights, mesh, weights)?,
+            3 => main_d::<3>(matches, edge_weights, mesh, weights)?,
+            n => anyhow::bail!("expected 2D or 3D mesh, got a {n}D mesh"),
+        }
     };
 
     let stdout = io::stdout();
```

Using the `--trace` argument of mesh-part, it is shown that the part that takes more time is this function invocation:
https://github.com/LIHPC-Computational-Geometry/coupe/blob/d531018d8fcc1906b77cb08fbe6075edef58da3d/src/algorithms/hilbert_curve.rs#L44-L45
and more specifically this parallel loop:
https://github.com/LIHPC-Computational-Geometry/coupe/blob/d531018d8fcc1906b77cb08fbe6075edef58da3d/src/algorithms/multi_jagged.rs#L276-L290

It turns out, the `scan` vector has less elements when not using the global threadpool, thus `ret` has a lot of duplicate entries and this loop does a lot of duplicate work.

### PR DESCRIPTION
**Pull request by hhirtz**
Tried to implement the 1DPART function from multijagged paper. It mostly works imbalance-wise, though there might be some edge cases that need to be covered and off-by-one errors to be fixed, I'm not 100% sure about the split indices.

### TODO

- [x] perf tests

Closes #201

### CODE REVIEW

**Comment by hhirtz** [comment]
Splitting a 3D mesh (30 million cells) into 64 parts:

- non-generic: https://github.com/LIHPC-Computational-Geometry/coupe/pull/240/commits/d409b6fe78bc675df5231514d6f47b4cf3056945
- generic: https://github.com/LIHPC-Computational-Geometry/coupe/pull/240/commits/8e9cd0556f2437081612d56db6f30c4ec0397ec3

generic impl incurs a 15-30% penalty, maybe because of (lack of) vectorisation.

non-generic impl has x2.5 speedup compared to master

base strong

**Comment by hhirtz** [comment]
base strong

It seems on high core counts, generic vs non-generic doesn't matter

**Comment by hhirtz** [comment]
Results on a 93-million-cell 3D mesh, split in 4096 parts really show the duplicate work bug in low core count runs:

base strong
